packages/arb-avm-cpp/data_storage/src/arbcore.cpp

```diff
@@ -953,73 +998,50 @@ rocksdb::Status ArbCore::reorgCheckpoints(
         std::cerr << "Reorg blockchain" << std::endl;
     }
 
+    if (last_machine) {
+        printMachineOutputInfo("Previous checkpoint",
+                               last_machine->machine_state.output);
+    }
+
     using checkpoint_pair =
         std::pair<MachineOutput, std::unique_ptr<MachineThread>>;
 
     auto found_checkpoint_or_status =
         [&]() -> std::variant<checkpoint_pair, rocksdb::Status> {
         ReadWriteTransaction tx(data_storage);
         auto checkpoint_it = tx.checkpointGetIterator();
+        auto reorg_machine_begin_time = std::chrono::steady_clock::now();
 
+        // Get last valid checkpoint from database
+        // (may or may not contain machine)
         auto output_or_status =
-            reorgToFirstMatchingCheckpoint(check_output, tx, checkpoint_it);
+            reorgToLastMatchingCheckpoint(check_output, tx, checkpoint_it);
         if (std::holds_alternative<rocksdb::Status>(output_or_status)) {
             return std::get<rocksdb::Status>(output_or_status);
         }
+        auto last_matching_database_checkpoint =
+            std::get<CheckpointVariant>(output_or_status);
 
-        // Save selected machine output, so we know how long to execute machine
-        // to match selected checkpoint.  This will be necessary if selected
-        // checkpoint does not include machine or if entry from machine cache is
-        // behind the selected checkpoint
-        MachineOutput selected_machine_output =
-            std::get<MachineOutput>(output_or_status);
-        std::cerr << "Selected checkpoint, total gas used: "
-                  << selected_machine_output.arb_gas_used
-                  << ", L1 block: " << selected_machine_output.l1_block_number
-                  << ", L2 block: " << selected_machine_output.l2_block_number
-                  << ", log count: " << selected_machine_output.log_count
-                  << ", messages count: "
-                  << selected_machine_output.fully_processed_inbox.count
-                  << ", timestamp: "
-                  << std::put_time(
-                         std::localtime((time_t*)&selected_machine_output
-                                            .last_inbox_timestamp),
-                         "%c")
-                  << "\n";
-
-        auto reorg_machine_begin_time = std::chrono::steady_clock::now();
-        auto nearest_machine_or_status = reorgToFirstMatchingMachineCheckpoint(
-            selected_machine_output, check_output, tx, checkpoint_it,
-            value_cache);
+        // Get last valid checkpoint with machine from database
+        auto last_matching_machine_checkpoint =
+            reorgToLastMatchingMachineCheckpoint(
```

Comment:
Maybe make this 
```cpp
auto last_matching_machine_or_status = loadLastMatchingMachine(
```
or something like that since it's not actually loading a checkpoint or deleting checkpoints (unless it happens to hit a corrupted one)

Reply:
:+1: